salt/roster/flat.py

```diff
@@ -44,19 +44,20 @@ def targets(tgt, tgt_type='glob', **kwargs):
     conditioned_raw = {}
     for minion in raw:
         conditioned_raw[str(minion)] = raw[minion]
-    rmatcher = RosterMatcher(conditioned_raw, tgt, tgt_type, 'ipv4')
+    rmatcher = RosterMatcher(conditioned_raw, tgt, tgt_type, 'ipv4', __opts__)
     return rmatcher.targets()
 
 
 class RosterMatcher(object):
     '''
     Matcher for the roster data structure
     '''
-    def __init__(self, raw, tgt, tgt_type, ipv='ipv4'):
+    def __init__(self, raw, tgt, tgt_type, ipv='ipv4', opts={}):
```

Comment:
I think you are mixing two different problems 1) re-usability and local variables 2) immutability.

* Right now the component _is_ accessing `__opts__` directly, so the component does not assume it is immutable. I don't see why I should change that.
* However, the component depends on a global, which means I can't use it in different contexts, which is what I intend to do.

So I still think

```python
class RosterMatcher(object):
    def __init__(self, raw, tgt, tgt_type, ipv='ipv4', opts={}):
    ...
```
and
```python
matcher = RosterMatcher(raw, tgt, tgt_type, 'ipv6', __opts__)
```
is the best option that solves that problem. If that is bad style for _some_ reason, then

```python
class RosterMatcher(object):
    def __init__(self, raw, tgt, tgt_type, ipv='ipv4', **kwargs):
    ...
```
and
```python
matcher = RosterMatcher(raw, tgt, tgt_type, 'ipv6', **__opts__)
```

Which I don't like because it does an implicit copy, which is unrelated to my problem.

I can just pass __opts__ as a kwargs itself like you describe above, without the copy.

```python
class RosterMatcher(object):
    def __init__(self, raw, tgt, tgt_type, ipv='ipv4', **kwargs):
        self.opts = kwargs.get('opts', {})
    ...
```
and
```python
matcher = RosterMatcher(raw, tgt, tgt_type, 'ipv6', opts=__opts__)
```
Which is the one I like more.